edxlearndot/learndot.py

```diff
@@ -0,0 +1,274 @@
+    """
+    pass
+
+
+def sort_enrolments_by_expiry(enrolment_list):
+    """
+    Sorts an array of Learndot enrolments by expiry date.
+
+    Arguments:
+        enrolment_list (list): a list of dictionaries representing
+            Learndot enrolments, as parsed from their API response.
+
+    Returns:
+        list: the input list, sorted by expiry date
+
+    Raises:
+        TypeError: if an expiry date is missing
+        ValueError: if an expiry date can't be parsed
+        OverflowError: if an expiry date can't be fit into the largest valid C integer
+    """
+    return sorted(enrolment_list, key=lambda e: dateutil.parser.parse(e.get("expiryDate")))
+
+
+class EnrolmentStatus(object):
+    """
+    Basically an enum of valid Learndot enrolment status values.
+
+    Provides the convenience method `is_valid`.
+    """
+    APPROVED = "APPROVED"
+    CANCELLED = "CANCELLED"
+    CONFIRMED = "CONFIRMED"
+    FAILED = "FAILED"
+    IN_PROGRESS = "IN_PROGRESS"
+    MISSED = "MISSED"
+    PASSED = "PASSED"
+    TENTATIVE = "TENTATIVE"
+
+    @classmethod
+    def is_valid(cls, status):
+        return hasattr(cls, status) and getattr(cls, status) == status
+
+
+class LearndotAPIClientBase(object):
+    """
+    Base class for clients of the Learndot API.
+    """
+
+    def get_api_key(self):
+        """
+        Returns the API key for the Learndot v2 API.
+        """
+        raise NotImplementedError
+
+    def get_api_base_url(self):
+        """
+        Returns the base URL for the Learndot v2 API.
+        """
+        raise NotImplementedError
+
+    def get_api_request_headers(self):
+        """
+        Returns the headers required for v2 API calls.
+        """
+        return {
+            "TrainingRocket-Authorization": self.get_api_key(),
+            "Content-Type": "application/json",
+            "Accept": "application/json"
+        }
+
+    def get_contact_search_url(self):
+        """
+        Returns the URL used to find contacts.
+        """
+        return os.path.join(self.get_api_base_url(), "api/rest/v2/manage/contact/search")
+
+    def get_enrolment_search_url(self):
+        """
+        Returns the URL used to find enrolments.
+        """
+        return os.path.join(self.get_api_base_url(), "api/rest/v2/manage/enrolment/search")
+
+    def get_enrolment_management_url_template(self):
+        """
+        Returns a template for the URL used to update enrolments.
+
+        The template URL contains a placeholder into which a numeric
+        enrolment ID should be substituted.
+        """
+        return os.path.join(
+            self.get_api_base_url(),
+            "api/rest/v2/manage/enrolment/{enrolment_id}"
+        )
+
+
+class LearndotAPIClient(LearndotAPIClientBase):
+    """
+    Client for the live Learndot API.
+    """
+
+    def get_api_key(self):
+        """
+        Returns the API key for the Learndot v2 API.
+        """
+        return settings.LEARNDOT_API_KEY
```

Comment:
When I only had the key in `lms.auth.json`, this was throwing an exception.
Not sure if this is just from how my devstack is setup, but I had to actually set it in `lms/env/common.py`.
Same error came with from the `settings.LEARNDOT_API_BASE_URL` below.

Reply:
Yeah, the JSON values would need to be pulled from the *TOKENS variables, which I hadn't done. I've changed the README to just say to set them, as I didn't think it was appropriate to add app-specific handling of that under lms/env.